# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
#Libraries
from functools import reduce
import numpy as np
import pandas as pd
from sqlalchemy import create_engine
import random
from random import choice
#import requests
#from bs4 import BeautifulSoup
import matplotlib.pyplot as plt
import seaborn as sns
from zipfile import ZipFile
#import wget
import gzip
import os
import locale


import sys
sys.executable

'C:\\Users\\ellio\\Anaconda3\\python.exe'

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [11]:
# your code here
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn import metrics
import numpy as np
import matplotlib.pyplot as plt

from sklearn import datasets
from sklearn.linear_model import LassoCV
from sklearn.linear_model import Lasso
from sklearn.model_selection import KFold
from sklearn.model_selection import GridSearchCV

diabetes = sklearn.datasets.load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [43]:
# your code here
#diabetes = sklearn.datasets.load_diabetes(return_X_y=False)
diabetes.keys()
pd.DataFrame(diabetes["feature_names"])

0
0  age
1  sex
2  bmi
3   bp
4   s1
5   s2
6   s3
7   s4
8   s5
9   s6

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [24]:
# your code here
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - Age
      - Sex
      - Body mass index
      - Average blood pressure
      - S1
      - S2
      - S3
      - S4
      - S5
      - S6

Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times `n_samples` (i.e. the sum of squares of each column totals 1).

Source URL:
https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html

For more information see:
Bra

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

# your answer here
How many attributes are there in the data? What do they mean?


What is the relation between diabetes['data'] and diabetes['target']?
Target: Column 11 is a quantitative measure of disease progression one year after baseline

How many records are there in the data?
442

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [46]:
# your code here
df_data = pd.DataFrame(diabetes["data"])
df_target = pd.DataFrame(diabetes["target"])
#df_data.shape #(442, 10)
#df_data.info() #10x float columns
#df_data.describe()
df_target.shape #(442, 1)
df_target.info()
df_target.describe()
df_data.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 442 entries, 0 to 441
Data columns (total 1 columns):
0    442 non-null float64
dtypes: float64(1)
memory usage: 3.5 KB


0         1         2         3         4         5         6  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

          7         8         9  
0 -0.002592  0.019908 -0.017646  
1 -0.039493 -0.068330 -0.092204  
2 -0.002592  0.002864 -0.025930  
3  0.034309  0.022692 -0.009362  
4 -0.002592 -0.031991 -0.046641

array([ 0.05068012, -0.04464164])

# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:


$y = \beta_0 + \beta_1·X_1 + \beta_2·X_2 + ... + \beta_n·X_n + \epsilon$


where:

- $X_1-X_n$: data 
- $\beta_0$: intercept 
- $\beta_1-\beta_n$: coefficients 
- $\epsilon$: error (cannot explained by model)
- $y$: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [72]:
# your code here
from  sklearn import linear_model 

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [73]:
# your code here
diabetes_model  = linear_model.LinearRegression()

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [80]:
# your code here
diabetes_data_train = df_data.iloc[:len(df_data)-20]
diabetes_target_train = df_target.iloc[:len(df_data)-20]
diabetes_data_test = df_data.iloc[len(df_data)-20:]
diabetes_target_test = df_target.iloc[len(df_data)-20:]

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [75]:
# your code here
diabetes_model.fit(diabetes_data_train, diabetes_target_train)
print(diabetes_model.fit(diabetes_data_train, diabetes_target_train).intercept_)
print(diabetes_model.fit(diabetes_data_train, diabetes_target_train).coef_)

[152.76430692]
[[ 3.03499549e-01 -2.37639315e+02  5.10530605e+02  3.27736980e+02
  -8.14131709e+02  4.92814588e+02  1.02848452e+02  1.84606489e+02
   7.43519617e+02  7.60951722e+01]]


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [78]:
# your code here
y_pred_test = diabetes_model.predict(diabetes_data_test)
y_pred_test


array([[197.61846908],
       [155.43979328],
       [172.88665147],
       [111.53537279],
       [164.80054784],
       [131.06954875],
       [259.12237761],
       [100.47935157],
       [117.0601052 ],
       [124.30503555],
       [218.36632793],
       [ 61.19831284],
       [132.25046751],
       [120.3332925 ],
       [ 52.54458691],
       [194.03798088],
       [102.57139702],
       [123.56604987],
       [211.0346317 ],
       [ 52.60335674]])

#### Print your `diabetes_target_test` and compare with the prediction. 

In [79]:
# your code here
print(diabetes_target_test)

         0
422  233.0
423   91.0
424  111.0
425  152.0
426  120.0
427   67.0
428  310.0
429   94.0
430  183.0
431   66.0
432  173.0
433   72.0
434   49.0
435   64.0
436   48.0
437  178.0
438  104.0
439  132.0
440  220.0
441   57.0


#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

In [84]:
# your answer here
diabetes_target_test - y_pred_test

0
422  35.381531
423 -64.439793
424 -61.886651
425  40.464627
426 -44.800548
427 -64.069549
428  50.877622
429  -6.479352
430  65.939895
431 -58.305036
432 -45.366328
433  10.801687
434 -83.250468
435 -56.333292
436  -4.544587
437 -16.037981
438   1.428603
439   8.433950
440   8.965368
441   4.396643

# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://onlinecourses.science.psu.edu/stat501/node/297/) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    $H_0 : \beta_1 = \beta_2 = ... = \beta_{10} = 0$

    $H_a$ : At least one $\beta_j \ne 0$ (for j = 1, 2, ..., 10)
    
    
* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

<div style="text-align:center"><img src="../images/statsmodels.png" width=600/></div>

In [87]:
# your code here
import statsmodels.api as sm
mod = sm.OLS(diabetes_target_train, diabetes_data_train)

res = mod.fit()

print(res.summary())

                                 OLS Regression Results                                
Dep. Variable:                      0   R-squared (uncentered):                   0.110
Model:                            OLS   Adj. R-squared (uncentered):              0.089
Method:                 Least Squares   F-statistic:                              5.109
Date:                Thu, 05 Dec 2019   Prob (F-statistic):                    4.77e-07
Time:                        15:46:39   Log-Likelihood:                         -2745.5
No. Observations:                 422   AIC:                                      5511.
Df Residuals:                     412   BIC:                                      5552.
Df Model:                          10                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

1. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

1. How will you modify your linear reguression model according to the test results above?

# your answer here
What is the F-score of your linear model and is the null hypothesis rejected?
5.109

Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?
??
How will you modify your linear reguression model according to the test results above?
??

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [94]:
# your code here
auto = pd.read_csv('../data/auto-mpg.csv')

Look at the first 5 rows using the `head()` function:

In [95]:
# your code here
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [96]:
# your code here
auto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 8 columns):
mpg             398 non-null float64
cylinders       398 non-null int64
displacement    398 non-null float64
horse_power     392 non-null float64
weight          398 non-null int64
acceleration    398 non-null float64
model_year      398 non-null int64
car_name        398 non-null object
dtypes: float64(4), int64(3), object(1)
memory usage: 25.0+ KB


In [97]:
auto.isna().sum()

mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

What is the newest model year and the oldest model year?

In [102]:
# your code here
# newest \t"chevy s-10"
# oldest \t"chevrolet chevelle malibu"

#auto.sort_values(by='model_year', ascending=False)
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Check the dataset for missing values and remove all rows containing at least one missing value.

In [105]:
# your code here
auto.dropna(inplace=True)
#auto.isna().sum()

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [107]:
# your code here
len(auto.cylinders.value_counts())

5

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [111]:
auto_X = auto[['cylinders','displacement','horse_power','weight','acceleration','model_year']]

In [112]:
# your code here
print(auto.shape)
#auto.drop('car_name', axis=1, inplace=True)
print(auto.shape)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(auto_X, auto.mpg, test_size=0.2)

(392, 7)
(392, 7)


Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [116]:
# your code here
auto_model  = linear_model.LinearRegression()

auto_model.fit(X_train, y_train)
print(auto_model.fit(X_train, y_train).intercept_)
print(auto_model.fit(X_train, y_train).coef_)

-16.474048494680517
[-0.11369192 -0.00175089  0.00966522 -0.00658771  0.02992202  0.77607771]



# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](https://en.wikipedia.org/wiki/Coefficient_of_determination). Its formula is:

<div style="text-align:center"><img src="../images/r-squared.png" width=300/></div>

where:

* yi is an actual data point.
* ŷi is the corresponding data point on the estimated regression line.

By adding the squares of the difference between all yi-ŷi pairs, we have a measure called SSE (*error sum of squares*) which is an application of the r squared score to indicate the extent to which the estimated regression model is different from the actual data. And we attribute that difference to the random error that is unavoidable in the real world. Obviously, we want the SSE value to be as small as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [117]:
# your code here
from sklearn.metrics import r2_score
y_pred = auto_model.predict(X_train)
y_pred
r2_score(y_train, y_pred)


0.8222072534876768

#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [120]:
# your code here
y_test_pred = auto_model.predict(X_test)

r2_score(y_test, y_test_pred)
#r2_score(y_test_pred, y_test)


0.7485996196312401

#### Explaining the results

Please, compare the results for the training and test sets and comment. What can you say?

In [ ]:
#your answer here
#The match is not as close

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [121]:
# your code here
X_train09, X_test09, y_train09, y_test09 = train_test_split(auto_X, auto.mpg, test_size=0.1)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [122]:
# your code here
auto_model09 = linear_model.LinearRegression()

auto_model09.fit(X_train09, y_train09)
print(auto_model.fit(X_train09, y_train09).intercept_)
print(auto_model.fit(X_train09, y_train09).coef_)

-13.47519908210202
[-0.39620363  0.00908182 -0.00404294 -0.00673384  0.01686417  0.75749642]


Compute the predicted values and r squared score for our new model and new sample data.

In [124]:
# your code here

y_train_pred09 = auto_model.predict(X_train09)

r2_score(y_train09, y_train_pred09)

0.8141040897604888

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [125]:
# your code here
y_test_pred09 = auto_model.predict(X_test09)

r2_score(y_test09, y_test_pred09)

0.7646376175895641

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE.

In [126]:
# Libraries
from sklearn.feature_selection import RFE

>>> from sklearn.datasets import make_friedman1
>>> from sklearn.feature_selection import RFE
>>> from sklearn.svm import SVR
>>> X, y = make_friedman1(n_samples=50, n_features=10, random_state=0)
>>> estimator = SVR(kernel="linear")
>>> selector = RFE(estimator, 5, step=1)
>>> selector = selector.fit(X, y)
>>> selector.support_
array([ True,  True,  True,  True,  True, False, False, False, False,
       False])
>>> selector.ranking_
array([1, 1, 1, 1, 1, 6, 4, 3, 2, 5])

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`.

In [128]:
# your code here
selector = RFE(auto_model, n_features_to_select=3)

Fit the model and print the ranking.

In [133]:
# your code here
selector = selector.fit(X_train09, y_train09)
selector.ranking_
auto_X.drop('horse_power', axis=1, inplace=True)

C:\Users\ellio\Anaconda3\lib\site-packages\pandas\core\frame.py:3940: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [137]:
# your code here
#X_train, X_test, y_train, y_test = train_test_split(auto_X, auto.mpg, test_size=0.2)
X_train_reduced, X_test_reduced, y_train_reduced, y_test_reduced = train_test_split(auto_X, auto.mpg, test_size=0.2)

Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [139]:
# your code here
auto_model_reduced = linear_model.LinearRegression()

auto_model_reduced.fit(X_train_reduced, y_train_reduced)
print(auto_model_reduced.fit(X_train_reduced, y_train_reduced).intercept_)
print(auto_model_reduced.fit(X_train_reduced, y_train_reduced).coef_)


mod = sm.OLS(y_train_reduced, X_train_reduced)

res = mod.fit()

print(res.summary())

-16.989456113365563
[-0.30749506  0.01332963 -0.00708161  0.17160157  0.76285618]
                                 OLS Regression Results                                
Dep. Variable:                    mpg   R-squared (uncentered):                   0.982
Model:                            OLS   Adj. R-squared (uncentered):              0.981
Method:                 Least Squares   F-statistic:                              3306.
Date:                Fri, 06 Dec 2019   Prob (F-statistic):                   3.42e-265
Time:                        07:50:25   Log-Likelihood:                         -821.41
No. Observations:                 313   AIC:                                      1653.
Df Residuals:                     308   BIC:                                      1672.
Df Model:                           5                                                  
Covariance Type:            nonrobust                                                  
                   coef    std err    

#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.